Sparkle/SUFileManager.m

```diff
@@ -100,29 +76,101 @@ - (instancetype)fileManagerByPreservingAuthorizationRights
+            *error = [NSError errorWithDomain:SUSparkleErrorDomain code:SUAuthenticationFailure userInfo:@{ NSLocalizedDescriptionKey: [NSString stringWithFormat:@"Failed creating authorization reference with status code %d.", createStatus] }];
         }
         _auth = NULL;
         return NO;
     }
+    
+    AuthorizationItem rightItems[] = {
+        // The right that allows us to run tools as root user
+        {.name = kAuthorizationRightExecute, .valueLength = 0, .value = NULL, .flags = 0}
+    };
+    
+    AuthorizationRights rights = {
+        .count = sizeof(rightItems) / sizeof(*rightItems),
+        .items = rightItems
+    };
+    
+    AuthorizationFlags flags =
+    (AuthorizationFlags)(kAuthorizationFlagDefaults | kAuthorizationFlagInteractionAllowed | kAuthorizationFlagExtendRights | kAuthorizationFlagPreAuthorize);
+    
+    // This will test if we can gain authorization for running utlities as root
+    OSStatus copyStatus = AuthorizationCopyRights(_auth, &rights, kAuthorizationEmptyEnvironment, flags, NULL);
+    if (copyStatus != errAuthorizationSuccess) {
+        if (error != NULL) {
+            if (copyStatus == errAuthorizationCanceled) {
+                *error = [NSError errorWithDomain:SUSparkleErrorDomain code:SUInstallationCancelledError userInfo:@{ NSLocalizedDescriptionKey: @"Authorization access was cancelled by the user." }];
+            } else {
+                *error = [NSError errorWithDomain:SUSparkleErrorDomain code:SUAuthenticationFailure userInfo:@{ NSLocalizedDescriptionKey: [NSString stringWithFormat:@"Failed granting authorization rights with status code %d.", copyStatus] }];
+            }
+        }
+        
+        AuthorizationFree(_auth, kAuthorizationFlagDefaults);
+        _auth = NULL;
+        
+        return NO;
+    }
+    
+    return YES;
+}
+
+- (BOOL)_authorizeAndExecutePath:(const char *)executablePath arguments:(char *const *)arguments error:(NSError * __autoreleasing *)error
+{
+    NSError *acquireError = nil;
+    if (![self _acquireAuthorizationWithError:&acquireError]) {
+        if (error != NULL) {
+            *error = acquireError;
+        }
+        return NO;
+    }
+    
+    FILE *pipe = NULL;
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
+    if (AuthorizationExecuteWithPrivileges(_auth, executablePath, kAuthorizationFlagDefaults, arguments, &pipe) != errAuthorizationSuccess) {
+#pragma clang diagnostic pop
+        if (error != NULL) {
+            *error = [NSError errorWithDomain:SUSparkleErrorDomain code:SUAuthenticationFailure userInfo:@{ NSLocalizedDescriptionKey:@"Failed to run authorization tool." }];
+        }
+        return NO;
+    }
+    
+    pid_t childPid = fcntl(fileno(pipe), F_GETOWN, 0);
+    int status = 0;
+    
+    pid_t waitResult;
+    do {
+        waitResult = waitpid(childPid, &status, 0);
+    } while (waitResult == -1 && errno == EINTR);
+    
+    fclose(pipe);
```

Comment:
Do you have to read from the pipe? With NSTask I've had tasks stuck forever, because their output buffer was full and were waiting for the parent to drain it.